drivers/bluetooth/hci/h4.c

```diff
@@ -159,15 +161,82 @@ static void reset_rx(void)
 
 static struct net_buf *get_rx(int timeout)
 {
+	struct net_buf *buf;
+
 	BT_DBG("type 0x%02x, evt 0x%02x", rx.type, rx.evt.evt);
 
 	if (rx.type == H4_EVT) {
-		return bt_buf_get_evt(rx.evt.evt, timeout);
+		buf = bt_buf_get_evt(rx.evt.evt, timeout);
+	} else {
+		buf = bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	}
+
+	if (buf) {
+		if (rx.discardable) {
+			buf->flags |= BUF_DISCARDABLE;
+		} else {
+			buf->flags &= ~BUF_DISCARDABLE;
+		}
 	}
 
-	return bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	return buf;
 }
 
+#if (CONFIG_BT_H4_DISCARD_RX_WAIT >= 0)
+static struct net_buf *get_rx_blocking(void)
+{
+	sys_slist_t list = SYS_SLIST_STATIC_INIT(&list);
+	bool discarded = false;
+	struct net_buf *buf;
+
+	/* If we have ACL flow control enabled then ACL buffers come from a
+	 * dedicated pool, and we cannot benefit from trying to discard
+	 * events from the RX queue.
+	 */
+	if (IS_ENABLED(CONFIG_BT_HCI_ACL_FLOW_CONTROL) && rx.type == H4_ACL) {
+		return get_rx(K_FOREVER);
+	}
+
+	buf = get_rx(CONFIG_BT_H4_DISCARD_RX_WAIT);
+	if (buf) {
+		return buf;
+	}
+
+	BT_WARN("Attempting to discard packets from RX queue");
+
+	while ((buf = net_buf_get(&rx.fifo, K_NO_WAIT))) {
+		if (!discarded && (buf->flags & BUF_DISCARDABLE)) {
+			net_buf_unref(buf);
+			discarded = true;
+		} else {
+			/* Since we use k_fifo_put_slist() instead of
+			 * net_buf APIs to insert back to the FIFO we
+			 * cannot support fragmented buffers. This should
+			 * never happen but better to "document" it in the
+			 * form of an assert here.
+			 */
+			__ASSERT(!(buf->flags & NET_BUF_FRAGS),
+				 "Fragmented buffers not supported");
+			sys_slist_append(&list, &buf->node);
+		}
+	}
+
+	if (!discarded) {
+		BT_WARN("Unable to discard anything from the RX queue");
+	}
+
+	/* Insert non-discarded packets back to the RX queue */
+	k_fifo_put_slist(&rx.fifo, &list);
```

Comment:
@andyross asking you since you seem to have touched the k_queue code most recently. The above code lines, starting with line 205, are iterating the contents of a k_fifo with the aim to remove one element from there. In this place of the code we have a guarantee that no other thread or ISR will attempt to access the k_fifo while we do this, however in the absence of any more suitable public APIs I'm having to remove each element one-by one and then re-insert the non-discarded elements using `k_fifo_put_slist()`. What I'd really want to do is the equivalent of `SYS_SFLIST_FOR_EACH_NODE()` and `sys_sflist_remove()` and stop iterating once a matching element is found and removed, however currently that requires accessing what I assume is private data of k_fifo (and k_queue). Would that still be ok, or any cleaner way to solve this?